torchtune/modules/transformer.py

```diff
@@ -206,6 +206,92 @@ def forward(self, tokens: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
         # shape: [b, s, d]
         h = self.norm(h)
 
-        # shape: [b, s, v]
         output = self.output(h).float()
         return output
+
+
+class TransformerClassifier(nn.Module):
```

Comment:
OK so if we are writing a standalone module I think this is the way to do it. The one question I would ask here is if we need a standalone module at all. To explain my thought process here a bit:

The only real difference between this class and an appropriately-parametrized builder of `TransformerDecoder` is the slicing you do to get just the EOS embedding (i.e. what you currently have in L282-296). But this is basically just "find the last non-padded index in the tokens and slice the logits based on that". In that case it may actually be conceptually simpler to do this in the training script. There you already know the padding index and it is not so different than our existing logic for e.g. ignore_index in our cross-entropy loss calculation. This way you have to pass the padding_index down into the model (in this case it is hardcoded, but in general I imagine we'd want it to be parametrized). If we handle this in the recipe, then you don't need to worry about this. You can also see in our PPO recipe where we do something similar in a [utility method](https://github.com/pytorch/torchtune/blob/bec7babec9c924a0ee7ad27e3f6582bc5bd1fef5/recipes/lora_dpo_single_device.py#L434-L440) with the cross entropy ignore index.

So I think the question comes down to what the full training script will look like. If I understand correctly, the outputs of this model will be used for training the reward model on a paired dataset, is that right? Do you think slicing to the EOS token makes more sense in the reward model training loop or in the model implementation? Admittedly it's kind of a philosophical question without a correct answer, but I bias towards doing it in the recipe because pad ID is a tokenizer property and the model should not have to be aware of it. But please let me know your thoughts here!

Reply:
Thanks for raising this. I think you make a good point - you could easily do this directly in the recipe. I think my main consideration would be (also philosophical) how explicit we want to be about that step. Generally speaking, any Seq2Label model will follow the step of classifying the last non-padding token, and produce a `[b, n_classes]` Tensor. Without the additional step of sequence pooling, a `mistral_classifier()` will return a `[b, s, n_classes]` Tensor.

You're right about what the training script looks like. Each recipe will need to replicate this step if we didn't use a dedicated module. If I was relatively new to all this, with good documentation I could understand by reading the recipe that you need to do some kind of pooling to convert `[b, s, n_classes]`, to `[b, n_classes]`, though it may not be clear that this is a common step that will occur after _any_ sequence classification recipe.

Looking forward, we'll want to pull the token padding ID directly from the tokenizer. We have a reference to this in the recipe. I think this is the most significant reason for me to agree that we should lose the separate module and just construct the classifier in the component builder, if we're okay with a `_classifier` model not _technically_ doing the complete classification (sans activation function).

Let me know your thoughts : )

Reply:
OK actually I am gonna raise an entirely orthogonal point here (sorry 😅). But during your description of dimensions it dawned on me that we are not necessarily doing things in the most efficient way here. We are applying the projection to the full sequence, when in reality we only need to do the projection on the EOS position. So rather than project and then slice we could slice and then project. This would decrease the size of the intermediate by a factor of `s`. I think the usefulness of this depends on how large the overall tensor is, if e.g. n_classes is really small then the intermediate may be small relative to other tensors in the model and the memory savings would not be as substantial. It also makes fitting into the existing class even thornier, since we would then be adding another operation before `TransformerDecoder`'s output projection.

Assuming that we don't follow the new wrench I threw in things, I think I am OK with either of the two approaches. Ultimately we won't know if we bump up into a wall until we integrate into a recipe, so feel free to make a call; then as we get closer to the actual training loop we can see how it works. How does that sound?

Reply:
Completely agree with all of this. Let’s not hack in changes for the memory savings now, I feel it makes most sense to start from a place of clean code and then see whether optimizations can/should be made after that. The utility method sounds good too, will take another pass once you update. Thanks for the patience on this discussion!